### ISSUE TITLE
TypeVar declared at class scope with generic bound specialized on an in-scope generic parameter TypeVar doesn't give an error

### ISSUE THREAD

**Issue by glyph** [python/mypy#15758]
**Bug Report**

**To Reproduce**

```python
from typing import TypeVar, Generic

T = TypeVar("T")

class A:
    pass

class B(A):
    pass

class G(Generic[T]):
    pass

anA = TypeVar("anA", bound=A)

class X(Generic[anA]):
    # error, as expected
    # AnInnerA = TypeVar("AnInnerA", bound=anA)
    AnInnerG = TypeVar("AnInnerG", bound=G[anA])

    def method(self, val: AnInnerG) -> None:
        pass

x: X[A] = X()
g: G[A] = G()

x.method(g)
```

[playground](https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=fae04e00e4632da34c5037f69f779953)

**Expected Behavior**

I would have expected to see an error on the line where `AnInnerG` was defined.

**Actual Behavior**

No error on `AnInnerG`'s definition, but I cannot find any type I can pass to `X.method` without provoking an error.

```
main.py:27: error: Value of type variable "AnInnerG" of "method" of "X" cannot be "G[A]"  [type-var]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.4.1
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files): n/a
- Python version used: 3.11

**Comment by erictraut** [comment]
The upper bound for a TypeVar cannot be generic. It must be concrete. This is spelled out in PEP 484. Since you've specified a bound of `G[anA]`, mypy should generate an error for this condition. I think there's another bug already logged for this false negative, but I'm unable to find it currently.

If you replace `bound=G[anA]` with `bound=G[Any]` or `bound=G`, your code will type check without errors.

It's not clear from your example why you're defining an inner type variable here. You could simply annotate the `val` parameter with `G[anA]` and eliminate the inner type variable. I'm guessing that your real use case is more complex than the sample shows.

**Comment by glyph** [comment]
> Since you've specified a bound of `G[anA]`, mypy should generate an error for this condition.

Yes, that's the bug. While I'd love to have higher-kinded types, I understand that they're not provided yet, the errors should just reflect that fact.

**Comment by glyph** [comment]
> I'm guessing that your real use case is more complex than the sample shows.

Yep, this is collateral damage on the way to trying to express a complex method-decorator constraint to Mypy; I'm still not entirely sure whether it's possible not.

**Comment by huix11** [comment]
12

**Comment by glyph** [comment]
It seems I am constantly trying to push this boundary no matter what I'm working on; I *think* this is a different manifestation of the same error-reporting bug:

```python3
from __future__ import annotations
from dataclasses import dataclass
from typing import Generic, TypeVar, reveal_type

T = TypeVar("T")
Vp = TypeVar("Vp")

@dataclass
class ListHaver(Generic[T, Vp]):
    t: T
    lt: Vp

    V = TypeVar("V", bound=list[T])

    @classmethod
    def new(cls, t: T, lt: V) -> ListHaver[T, V]:
        return ListHaver(t, lt)

class A:
    pass

class Special(list[A]):
    def special(self) -> None:
        print("special")

anything: list[A] = Special()
reveal_type(ListHaver.new(A(), anything))
```
